compiler/rustc_middle/src/mir/interpret/allocation.rs

```diff
@@ -563,22 +563,12 @@ impl<Tag, Extra> Allocation<Tag, Extra> {
 
         let mut ranges = smallvec::SmallVec::<[u64; 1]>::new();
         let initial = self.init_mask.get(range.start);
-        let mut cur_len = 1;
-        let mut cur = initial;
 
-        for i in 1..range.size.bytes() {
-            // FIXME: optimize to bitshift the current uninitialized block's bits and read the top bit.
-            if self.init_mask.get(range.start + Size::from_bytes(i)) == cur {
-                cur_len += 1;
-            } else {
-                ranges.push(cur_len);
-                cur_len = 1;
-                cur = !cur;
-            }
+        for chunk in self.init_mask.range_as_init_chunks(range.start, range.end()) {
+            let len = chunk.range().end.bytes() - chunk.range().start.bytes();
```

Comment:
This seems to make the assumption that the iterator will always alternate between `InitChunk::Init` and `InitChunk::Uninit`, with the first element matching the value at `range.start`. We should at least clearly document this as part of the contract of `range_as_init_chunks`.